File: common/data/src/main/java/de/christian2003/smarthome/data/model/cert/CertHandler.java
package de.christian2003.smarthome.data.model.cert;

import android.content.Context;
import android.net.Uri;
import android.util.Log;
import androidx.annotation.NonNull;
import java.io.File;
import java.io.InputStream;
import java.nio.file.Files;
import java.security.KeyStore;
import java.security.SecureRandom;
import javax.net.ssl.KeyManagerFactory;
import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManagerFactory;


/**
 * Class implements a handler that can save a client certificate (and password) through
 * {@link #importNewCertFromUri(Uri, String)}. Get an {@linkplain SSLContext} for all requests through
 * {@link #getSSLContext()} which includes the client certificate.
 */
public class CertHandler {

    /**
     * Field stores the tag to use for logging.
     */
    private static final String TAG = "CertHandler";


    /**
     * Attribute stores the key store helper.
     */
    private final KeyStoreHelper helper;


    /**
     * Constructor instantiates a new cert handler.
     *
     * @param context   Context for the cert handler.
     */
    public CertHandler(Context context) {
        helper = new KeyStoreHelper(context);
    }


    /**
     * Method copies the .pfx certificate of the specified URI to the internal storage and saves
     * the provided password for the certificate.
     *
     * @param uri           URI to the .pfx-file to copy to internal storage.
     * @param certPassword  The password of the certificate.
     * @throws Exception    Some error occurred.
     */
    public void importNewCertFromUri(@NonNull Uri uri, String certPassword) throws Exception {
        helper.saveCertPassword(certPassword);
        helper.saveCertToInternalStorage(uri);
    }


    /**
     * Method returns the SSLContext to use for connections using the custom client certificate. If
     * no client certificate is available or any other error occurs, an exception is thrown.
     *
     * @return              SSLContext with the client certificate to use with requests to the server.
     * @throws Exception    Some error occurred.
     */
    public SSLContext getSSLContext() throws Exception {
        File certFile = helper.getCertFile();
        String password = helper.getCertPassword();

        if (!certFile.exists() || password == null) {
            Log.e(TAG, "Certificate or password not available");
            throw new IllegalStateException("Certificate or password not available");
        }

        try (InputStream certStream = Files.newInputStream(certFile.toPath())) {
            KeyStore keyStore = KeyStore.getInstance("PKCS12");
            keyStore.load(certStream, password.toCharArray());

            KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance("X509");
            keyManagerFactory.init(keyStore, password.toCharArray());

            TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance("X509");
            trustManagerFactory.init((KeyStore)null); //I have no idea why I have to cast "null" to "KeyStore"...

            SSLContext sslContext = SSLContext.getInstance("TLS");
            sslContext.init(keyManagerFactory.getKeyManagers(), trustManagerFactory.getTrustManagers(), new SecureRandom());

            return sslContext;
        }
        catch (Exception e) {
            Log.e(TAG, e.getMessage() != null ? e.getMessage() : "Unknown error occurred");
            throw e;
        }
    }


    /**
     * Method removes the certificate and the corresponding password from the app.
     *
     * @throws Exception    The file cannot be deleted.
     */
    public void removeCert() throws Exception {
        helper.removeCertPassword();
        File certFile = helper.getCertFile();
        boolean removed;
        if (certFile.exists()) {
            try {
                removed = certFile.delete();
            }
            catch (Exception e) {
                Log.e(TAG, e.getMessage() != null ? e.getMessage() : "Unknown error occurred");
                throw e;
            }
            if (!removed) {
                Log.e(TAG, "Cannot remove certificate file");
                throw new Exception("Cannot remove certificate file");
            }
        }
    }

}

File: common/data/src/main/java/de/christian2003/smarthome/data/model/cert/CertHandler.kt
package de.christian2003.smarthome.data.model.cert

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import java.security.MessageDigest
import java.security.cert.X509Certificate
import javax.net.ssl.SSLContext


class CertHandler(

    /**
     * Attribute stores the context to use for all operations.
     */
    val context: Context

) {

    /**
     * Attribute stores the preferences in which the certificate alias is stored.
     */
    private val preferences = context.getSharedPreferences("smart_home", Context.MODE_PRIVATE)


    fun validateCert(cert: X509Certificate): SslTrustResponse {
        if (preferences.getBoolean("unsafe_cert_validation", false)) {
            return SslTrustResponse(SslTrustStatus.Trusted, cert) //Skip validation if unsafe cert validation is enabled.
        }
        val fingerprint = getCertFingerprint(cert)
        return if (isCertTrusted(fingerprint)) {
            SslTrustResponse(SslTrustStatus.Trusted, cert)
        } else {
            SslTrustResponse(SslTrustStatus.Untrusted, cert)
        }
    }


    fun saveCert(cert: X509Certificate) {
        val trustedCerts: MutableSet<String> = preferences.getStringSet("trusted_certs", emptySet())!!.toMutableSet()
        trustedCerts.add(getCertFingerprint(cert))
        val editor: SharedPreferences.Editor = preferences.edit()
        editor.putStringSet("trusted_certs", trustedCerts)
        editor.apply()
    }


    /**
     * Method returns the SSL context to use for client authentication. This can return null if no
     * certificate has been selected from the Android key store.
     *
     * @return  SSL context to use for client authentication.
     */
    fun getSSLContext(): SSLContext? {
        try {
            val alias = preferences.getString("cert_alias", null) ?: return null
            val keyManager = KeyChainKeyManager(context, alias)

            val sslContext = SSLContext.getInstance("TLS")
            sslContext.init(arrayOf(keyManager), null, null)
            return sslContext
        } catch (e: Exception) {
            Log.e("Certificates", "Cannot create SSLContext object: ${e.message}")
        }
        return null


        /*
        try {
            val alias: String? = preferences.getString("cert_alias", null)
            if (alias != null) {
                val key: PrivateKey? = KeyChain.getPrivateKey(context, alias)
                if (key != null) {
                    val chain: Array<X509Certificate>? = KeyChain.getCertificateChain(context, alias)
                    if (chain != null) {
                        val keyStore: KeyStore? = KeyStore.getInstance("PKCS12")
                        keyStore?.load(null, null)
                        keyStore?.setKeyEntry(alias, key, null, chain)

                        val keyManagerFactory: KeyManagerFactory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm())
                        keyManagerFactory.init(keyStore, null)

                        val trustManagerFactory: TrustManagerFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())
                        trustManagerFactory.init(null as KeyStore?)

                        val sslContext: SSLContext = SSLContext.getInstance("TLS")
                        sslContext.init(keyManagerFactory.keyManagers, trustManagerFactory.trustManagers, null)
                        return sslContext
                    }
                }
            }
        }
        catch (e: Exception) {
            val message: String = if (e.message == null) { "Unknown error" } else { e.message!! }
            Log.e("Certificates", "Cannot create SSLContext object: $message")
        }

        return null
        */
    }


    private fun isCertTrusted(fingerprint: String): Boolean {
        val trustedCerts: Set<String>? = preferences.getStringSet("trusted_certs", emptySet())
        trustedCerts!!.forEach { cert ->
            Log.d("CertHandler", "Trusted Fingerprint: $cert")
        }
        Log.d("CertHandler", "Cert Fingerprint: $fingerprint")
        return trustedCerts!!.contains(fingerprint)
    }


    private fun getCertFingerprint(cert: X509Certificate): String {
        try {
            val md = MessageDigest.getInstance("SHA-256")
            val publicKey = md.digest(cert.encoded)

            val hexString = StringBuilder()
            for (b in publicKey) {
                hexString.append(String.format("%02X:", b))
            }
            return hexString.substring(0, hexString.length - 1)
        }
        catch (e: Exception) {
            return ""
        }
    }

}
